Add duplicate accounts backend feature including api call to append, remove and get duplicated users

pcolt opened the pull request:
This partially closes #4355.

- It adds a new many-to-many linking table to link duplicated users together (therefore includes a migration). The table has been designed to easily support in the future other kinds of links between users a part of the "duplicated" type (just adding a new value in the enumerator `UserLinkType`).
- It adds three new endpoints to link together two users at a time, remove a link and get all users linked to one given user.
- What I left out is the part regarding the `_user_to_details` function that I think was mentioned in the ticket. It was not easy and clear to me how to implement this part and in the ticket was suggested to leave this part for a following issue.

No specific steps required to review. I have mainly used unit tests to check everything.

Being new to the project and to some of the technologies used I obviously open to get any suggestion for change.

**Backend checklist**
- [x] Formatted my code by running `ruff check --select I --fix . && ruff check . --fix && ruff format .` in `app/backend`
- [x] Added tests for any new code or added a regression test if fixing a bug
- [x] All tests pass
- [x] Run the backend locally and it works
- [x] Added migrations if there are any database changes, rebased onto `develop` if necessary for linear migration history

pcolt committed:
add table, service, tests to link two duplicate users

pcolt committed:
fix exception link already exist + formatting code

pcolt committed:
add remove link service, protos, tests

pcolt committed:
add get link duplicated service, protos, tests

pcolt committed:
remove link user type travel buddy and flat mates

pcolt committed:
Merge branch 'develop' into backend/feature/duplicate-accounts

aapeliv commented on app/backend/src/couchers/models.py:
```diff
@@ -2667,3 +2667,32 @@ class AccountDeletionReason(Base):
+    __tablename__ = "user_links"
+
+    id = Column(BigInteger, primary_key=True)
+    user1_id = Column(ForeignKey("users.id"), nullable=False)
+    user2_id = Column(ForeignKey("users.id"), nullable=False)
```
Add `index=True` to automatically create a lookup index for these

aapeliv commented on app/backend/src/couchers/servicers/admin.py:
```diff
@@ -578,3 +581,76 @@ def EditReply(self, request, context, session):
+            link_type=UserLinkType.duplicate_account,
+        )
+
+        try:
+            session.add(user_link)
+            session.commit()
+        except exc.IntegrityError:
+            context.abort(grpc.StatusCode.ALREADY_EXISTS, errors.USER_LINK_ALREADY_EXISTS)
```
Switch this to a lookup: `IntegrityError` will catch other stuff too.

aapeliv commented on app/proto/admin.proto:
```diff
@@ -93,6 +99,8 @@ message UserDetails {
   int64 pending_mod_notes_count = 16;
   int64 acknowledged_mod_notes_count =17;
+
+  repeated UserLink links = 18;
```
Please use `user_links`

pcolt replied:
Oh, my apologies I forgot to revert this. I was trying to add the duplicate links to the _user_to_details administration function, but I got a few errors and wasn't sure what was the right way to implement this.

aapeliv commented on app/proto/admin.proto:
```diff
@@ -296,3 +304,32 @@ message EditReplyReq {
+}
+
+message LinkUsersDuplicatedRes {
+    int64 link_id = 1;
```
`user_link_id`, etc

pcolt replied:
Not 100% sure about this, cause I see this a bit of a repetition. Anyway I changed all `link_id` to `user_link_id` and all `link_type` to `user_link_type` in the protos and in the consequent requests/responses but I didn't change the models/column tables. Hope this was what you intended?

pcolt committed:
PR fixes: add indexes, fix catching error, rename protos

pcolt committed:
Merge branch 'develop' into backend/feature/duplicate-accounts

pcolt committed:
new tables, logic and tests to group duplicated users

pcolt committed:
format

pcolt committed:
prepare file admin.py for merge

pcolt committed:
Merge branch 'develop' into backend/feature/duplicate-accounts

pcolt committed:
move duplicate users group back as before merging

pcolt committed:
add migration after merging develop

pcolt committed:
cleaning and commenting

pcolt committed:
last commit

pcolt commented:
Following Aapeli (and Jesse) suggestions 

> "The one thing that was discussed earlier was about having 'groups' of linked users, instead of just 1-to-1 links. So e.g. with dupe accounts you'll often see people make many accounts, and then it'd be hard to link each one together." 

I changed a bit the architecture adding a new association table `UserGroupMember` which groups users together instead of linking them one by one. 

The only downside of this approach is that a check before grouping has to be made ensuring that there are not two or more UserGroupMember associated with same group type (to avoid having repetitions of duplicated groups that should be unique instead).

I also added some methods to the sqlachemy models like `has_user` and `get_group_duplicated` that I honestly never used before but they were somehow a pattern used in the codebase already.

An UML of the new tables using plantuml:

![Image](https://github.com/user-attachments/assets/3c36e34b-865c-4f10-befc-b25a05b7fda6)

- [x] Formatted my code by running ruff check --select I --fix . && ruff check . --fix && ruff format . in app/backend
- [x] Added tests for any new code or added a regression test if fixing a bug
- [x] All tests pass
- [x] Run the backend locally and it works
- [x] Added migrations if there are any database changes, rebased onto develop if necessary for linear migration history

aapeliv commented on app/backend/src/couchers/models.py:
```diff
@@ -2667,3 +2673,37 @@ class AccountDeletionReason(Base):
+    duplicate_account = enum.auto()
+
+
+class UserGroup(Base):
```
Let's specialize this and drop the `UserGroupType`, instead calling this `ModerationUserList` and do it just for duplciate users. "Group" is too overloaded here, since it's a concept in our communities side as well.

pcolt replied:
done

aapeliv commented on app/backend/src/couchers/models.py:
```diff
@@ -2667,3 +2673,37 @@ class AccountDeletionReason(Base):
+    __tablename__ = "user_group_members"
+
+    user_id = Column(ForeignKey("users.id"), primary_key=True)
+    group_id = Column(ForeignKey("user_groups.id"), primary_key=True)
```
You should add a database constraint here to make sure each user is in each group at most once.

I see you implemented a restriction of one user can be in at most one group. You can ask Jesse what his preference is: in my view it might be useful to allow mods to have multiple user lists and use them for whatever is needed: e.g. duplicate accounts, but also to link them together otherwise (if they are related in some other activity, etc).

For each user being able to be in at most one group, you can add:

```py
    __table_args__ = (UniqueConstraint("user_id"),)
```

For each user being able to be listed in each group at most once:

```py
    __table_args__ = (UniqueConstraint("user_id", "group_id"),)
```

jesseallhands replied:
>I see you implemented a restriction of one user can be in at most one group. You can ask Jesse what his preference is: in my view it might be useful to allow mods to have multiple user lists and use them for whatever is needed: e.g. duplicate accounts, but also to link them together otherwise (if they are related in some other activity, etc).

I agree with allowing mods to have multiple user lists

pcolt replied:
done

aapeliv commented on app/backend/src/couchers/servicers/admin.py:
```diff
@@ -584,6 +586,77 @@ def EditReply(self, request, context, session):
+            if user not in user_group_duplicated.users:
+                user_group_duplicated.users.append(user)
+
+        session.commit()
```
No need for a commit here: it autocommits when it leaves the `session_scope` contextmanager (defined in `interceptors.py:310` (see `SessionInterceptor`)

pcolt replied:
OK, done

aapeliv commented on app/backend/src/couchers/servicers/admin.py:
```diff
@@ -584,6 +586,77 @@ def EditReply(self, request, context, session):
+        if not duplicate_group:
+            context.abort(grpc.StatusCode.FAILED_PRECONDITION, errors.USER_NOT_IN_ANY_DUPLICATE_GROUP)
+
+        duplicate_group.users.remove(user)
```
If it's the last user in the group, then we should remove the group too, right?

pcolt replied:
yes you are right. Done

aapeliv commented on app/backend/src/couchers/models.py:
```diff
@@ -427,6 +429,10 @@ def phone_code_expired(self):
 
+    def get_group_duplicated(self):
+        """Returns the duplicate account group if user belongs to one, otherwise returns None."""
+        return next((group for group in self.groups if group.group_type == UserGroupType.duplicate_account), None)
```
This is better to move into the code where it runs.

pcolt replied:
done

aapeliv commented on app/backend/src/couchers/models.py:
```diff
@@ -2667,3 +2673,37 @@ class AccountDeletionReason(Base):
+    # Relationships
+    users = relationship("User", secondary="user_group_members", back_populates="groups")
+
+    def has_user(self, user):
+        """Check if user is in group"""
+        return user.id in {u.id for u in self.users}
```
This kind of check is better done in SQL. Right now when you call this function, it will pull the set of users for this group into python, construct the set, then check membership. With a SQL index, you can do that in very little computation on the database side.

pcolt replied:
I moved it where it is used in the tests' code

pcolt committed:
remove last migration

pcolt committed:
Merge branch 'develop' into backend/feature/duplicate-accounts

pcolt committed:
#5967 moderation list rename and fixes following pr review

pcolt committed:
use faster set comprehensions

pcolt commented:
This time I created only two endpoints: one for adding and another for removing users from the new moderation lists. I thought it was better to do one thing at a time. Also because the part of getting the lists is not yet clear to me and in the original issue was stated that @aapeliv would manage it. Of course I am open to implement it afterwards if needed

aapeliv reviewed (commented):
I think what's here looks good! Just one more thing is that I think we need a `ListUserModerationLists` or similar to get the moderation list the user is in.

Commit:
Merge branch 'develop' into backend/feature/duplicate-accounts

Commit:
add endpoint to get list of moderation lists a user is in

aapeliv commented on app/proto/admin.proto:
```diff
@@ -353,6 +355,15 @@ message AddUsersToModerationUserListReq {
+}
+
+message ListModerationUserListsRes {
+  repeated int64 moderation_list_ids = 1;
```
Maybe make a

```proto3
message ModerationList {
  int64 moderation_list_id = 1;
  repeated uint32 member_ids = 2;
}
```

and return repeated objects like that?

pcolt replied:
This is done. I have now to undo the migration, merge develop and reapply the migrations. @aapeliv may you review this in the meantime?

Commit:
add list of members in list of moderation lists response

jesseallhands commented:
@aapeliv did you finish your review on this?

aapeliv reviewed (dismissed):
LGTM

aapeliv committed:
Merge branch 'develop' into backend/feature/duplicate-accounts
